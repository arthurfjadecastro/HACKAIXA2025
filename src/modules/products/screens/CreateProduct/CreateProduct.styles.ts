import { StyleSheet } from 'react-native';
import { colors } from '@/design-system/tokens';

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.surface.background,
  },
  
  // Header
  header: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingHorizontal: 20, // Espaçamento lateral otimizado
    paddingTop: 16, // Padding top reduzido
    paddingBottom: 20, // Padding bottom aumentado para mais respiro
    backgroundColor: colors.surface.background,
    borderBottomWidth: 1,
    borderBottomColor: colors.border.light,
    minHeight: 64, // Altura mínima para melhor proporção
  },
  backButton: {
    padding: 8, // Padding reduzido
    marginRight: 16, // Margem direita otimizada
    width: 40,
    height: 40,
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 20, // Cantos arredondados para melhor visual
  },
  title: {
    flex: 1,
    color: colors.primary.main,
    fontSize: 18, // Fonte ligeiramente menor para melhor proporção
    fontWeight: '600',
    lineHeight: 24, // Line height para melhor legibilidade
  },
  
  // Content
  keyboardView: {
    flex: 1,
  },
  content: {
    flex: 1,
  },
  contentContainer: {
    paddingBottom: 32, // Padding bottom aumentado
  },
  form: {
    paddingHorizontal: 0, // Removido - padding será aplicado no FormContainer
    paddingTop: 24, // Padding top otimizado
    gap: 24, // Gap aumentado entre seções
  },
  
  // Footer
  footer: {
    padding: 20, // Padding consistente
    backgroundColor: colors.surface.background,
    borderTopWidth: 1,
    borderTopColor: colors.border.light,
    elevation: 4, // Sombra no Android
    shadowColor: '#000', // Sombra no iOS
    shadowOffset: {
      width: 0,
      height: -2,
    },
    shadowOpacity: 0.1,
    shadowRadius: 3.84,
  },
  submitButton: {
    width: '100%',
    minHeight: 48, // Altura mínima aumentada
    borderRadius: 12, // Cantos arredondados
  },
  submitButtonEnabled: {
    backgroundColor: colors.brand.orange.primary,
  },
  submitButtonDisabled: {
    backgroundColor: colors.action.disabled.bg,
  },
});
