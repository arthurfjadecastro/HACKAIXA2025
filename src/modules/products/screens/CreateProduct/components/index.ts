export { CreateProductHeader } from './CreateProductHeader';
export { CreateProductForm } from './CreateProductForm';
export { CreateProductFooter } from './CreateProductFooter';
