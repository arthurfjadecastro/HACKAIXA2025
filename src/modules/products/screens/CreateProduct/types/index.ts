export interface FormData {
  // Seleção de categoria/subtipo
  categoria: 'CONSIGNADO' | 'HABITACAO' | 'OUTRO' | '';
  subtipo: 'INSS' | 'CONVENIO' | 'CLT_SUSPENSO' | 'N/A' | '';
  convenio_id?: string;
  convenio_selected?: string; // Para seleção simplificada (militar, funcef, tjdft)
  
  // Campos originais
  name: string;
  interestRate: string;
  maxTerm: string;
  normative: string;
  
  // Campos preenchidos automaticamente - Consignado
  prazo_minimo?: number;
  prazo_maximo?: number;
  margem_consignavel?: number;
  canais_permitidos?: string[];
  taxa_faixa_a_concessao?: number;
  taxa_faixa_a_renovacao?: number;
  taxa_faixa_b_concessao?: number;
  taxa_faixa_b_renovacao?: number;
  taxa_faixa_c_concessao?: number;
  taxa_faixa_c_renovacao?: number;
  portabilidade_prospeccao_min?: number;
  portabilidade_prospeccao_max?: number;
  exige_validacao_remota?: boolean;
  exige_validacao_presencial?: boolean;
  exige_certificacao_digital?: boolean;
  
  // Campos específicos de Habitação
  sistema_amortizacao?: 'SAC' | 'PRICE' | '';
  ltv_max_percentual?: number;
  entrada_min_percentual?: number;
  indexadores_permitidos?: Array<{id: string, descricao: string}>;
  seguros_obrigatorios?: Array<{id: string, descricao: string}>;
  
  // Campos específicos de OUTRO
  subcategoria_outro?: string;
  valor_min?: number;
  valor_max?: number;
  prazo_min_meses?: number;
  prazo_max_meses?: number;
  quantidade_min?: number;
  quantidade_max?: number;
  taxa_percentual?: number;
  taxa_valor_fixo?: number;
  taxa_por_unidade?: number;
  taxa_observacao?: string;
  canais_selecionados?: string[];
  faixa_etaria_min?: number;
  faixa_etaria_max?: number;
  
  // Campos comuns
  observacoes?: string[];
  fonte_dados?: string;
}

export interface FormErrors {
  categoria?: string;
  subtipo?: string;
  convenio_id?: string;
  name?: string;
  interestRate?: string;
  maxTerm?: string;
  normative?: string;
}

export interface FieldValidation {
  isValid: boolean;
  error?: string;
}

export type FieldName = keyof FormData;

export interface UseFormReturn {
  formData: FormData;
  errors: FormErrors;
  touched: Record<string, boolean>;
  isLoading: boolean;
  isFormValid: boolean;
  updateField: (fieldName: FieldName, value: string | number | boolean | string[]) => void;
  handleBlur: (fieldName: FieldName) => void;
  handleSubmit: () => Promise<void>;
  scrollToFirstError: () => void;
}
