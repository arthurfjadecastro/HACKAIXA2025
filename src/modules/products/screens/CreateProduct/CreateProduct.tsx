import React from 'react';
import { ScrollView } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';

import { CreateProductHeader, CreateProductForm, CreateProductFooter } from './components';
import { useCreateProductForm } from './hooks/useCreateProductFormSimple';
import { styles } from './CreateProduct.styles';

const CreateProduct: React.FC = () => {
  const {
    formData,
    isLoading,
    isFormValid,
    updateField,
    handleSubmit,
    scrollViewRef,
    isConvenioAlreadyRegistered,
  } = useCreateProductForm();

  return (
    <SafeAreaView style={styles.container}>
      <CreateProductHeader />
      
      <ScrollView
        ref={scrollViewRef}
        style={styles.content}
        contentContainerStyle={styles.contentContainer}
        showsVerticalScrollIndicator={false}
        keyboardShouldPersistTaps="handled"
      >
        <CreateProductForm
          formData={formData}
          updateField={updateField}
          isConvenioAlreadyRegistered={isConvenioAlreadyRegistered}
        />
      </ScrollView>

      <CreateProductFooter
        isLoading={isLoading}
        isFormValid={isFormValid}
        onSubmit={handleSubmit}
      />
    </SafeAreaView>
  );
};

export default CreateProduct;