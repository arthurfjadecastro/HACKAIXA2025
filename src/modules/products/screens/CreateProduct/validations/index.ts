import { FormData, FieldName, FieldValidation } from '../types';
import { 
  validateName, 
  validateInterestRate, 
  validateMaxTerm, 
  validateNormative,
  validateCategoria,
  validateSubtipo
} from './fieldValidations';

export const validateField = (fieldName: FieldName, value: string): FieldValidation => {
  switch (fieldName) {
    case 'name':
      return validateName(value);
    case 'interestRate':
      return validateInterestRate(value);
    case 'maxTerm':
      return validateMaxTerm(value);
    case 'normative':
      return validateNormative(value);
    case 'categoria':
      return validateCategoria(value);
    case 'subtipo':
      return validateSubtipo(value);
    default:
      return { isValid: true };
  }
};

export const validateForm = (formData: FormData): boolean => {
  // Validações básicas sempre necessárias
  const categoriaValid = validateCategoria(formData.categoria).isValid;
  const subtipoValid = validateSubtipo(formData.subtipo).isValid;
  
  // Se categoria e subtipo estão selecionados, validar campos específicos
  if (categoriaValid && subtipoValid && formData.categoria && formData.subtipo) {
    const nameValid = validateName(formData.name).isValid;
    const normativeValid = validateNormative(formData.normative).isValid;
    
    return nameValid && normativeValid;
  }
  
  return categoriaValid && subtipoValid;
};

export const validateAllFields = (formData: FormData) => {
  const errors: Partial<Record<FieldName, string>> = {};
  
  // Validar categoria e subtipo primeiro
  const categoriaValidation = validateCategoria(formData.categoria);
  if (!categoriaValidation.isValid && categoriaValidation.error) {
    errors.categoria = categoriaValidation.error;
  }
  
  const subtipoValidation = validateSubtipo(formData.subtipo);
  if (!subtipoValidation.isValid && subtipoValidation.error) {
    errors.subtipo = subtipoValidation.error;
  }
  
  // Se categoria e subtipo são válidos e estão preenchidos, validar outros campos
  if (categoriaValidation.isValid && subtipoValidation.isValid && 
      formData.categoria && formData.subtipo) {
    
    const fieldOrder: FieldName[] = ['name', 'normative'];
    
    fieldOrder.forEach(fieldName => {
      const fieldValue = formData[fieldName];
      if (typeof fieldValue === 'string') {
        const validation = validateField(fieldName, fieldValue);
        if (!validation.isValid && validation.error) {
          errors[fieldName] = validation.error;
        }
      }
    });
  }

  return errors;
};
