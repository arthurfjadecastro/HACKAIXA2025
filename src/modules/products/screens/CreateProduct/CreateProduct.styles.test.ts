import { styles } from './CreateProduct.styles';

describe('CreateProduct Styles', () => {
  it('should export styles object', () => {
    expect(styles).toBeDefined();
    expect(typeof styles).toBe('object');
  });

  it('should have container style', () => {
    expect(styles.container).toBeDefined();
    expect(styles.container.flex).toBe(1);
  });

  it('should have header style', () => {
    expect(styles.header).toBeDefined();
    expect(styles.header.flexDirection).toBe('row');
    expect(styles.header.alignItems).toBe('center');
    expect(styles.header.paddingHorizontal).toBe(20);
    expect(styles.header.paddingTop).toBe(16);
    expect(styles.header.paddingBottom).toBe(20);
    expect(styles.header.minHeight).toBe(64);
  });

  it('should have backButton style', () => {
    expect(styles.backButton).toBeDefined();
    expect(styles.backButton.padding).toBe(8);
    expect(styles.backButton.marginLeft).toBe(-8);
    expect(styles.backButton.marginRight).toBe(16);
    expect(styles.backButton.width).toBe(40);
    expect(styles.backButton.height).toBe(40);
    expect(styles.backButton.justifyContent).toBe('center');
    expect(styles.backButton.alignItems).toBe('center');
    expect(styles.backButton.borderRadius).toBe(20);
  });

  it('should match snapshot', () => {
    expect(styles).toMatchSnapshot();
  });
});
