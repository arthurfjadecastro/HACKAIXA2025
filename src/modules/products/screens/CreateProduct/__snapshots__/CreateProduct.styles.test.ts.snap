// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateProduct Styles should match snapshot 1`] = `
{
  "backButton": {
    "alignItems": "center",
    "borderRadius": 20,
    "height": 40,
    "justifyContent": "center",
    "marginLeft": -8,
    "marginRight": 16,
    "padding": 8,
    "width": 40,
  },
  "container": {
    "backgroundColor": "#FFFFFF",
    "flex": 1,
  },
  "content": {
    "flex": 1,
  },
  "contentContainer": {
    "paddingBottom": 32,
  },
  "footer": {
    "backgroundColor": "#FFFFFF",
    "borderTopColor": "#EEEEEE",
    "borderTopWidth": 1,
    "elevation": 4,
    "padding": 20,
    "shadowColor": "#000",
    "shadowOffset": {
      "height": -2,
      "width": 0,
    },
    "shadowOpacity": 0.1,
    "shadowRadius": 3.84,
  },
  "form": {
    "gap": 24,
    "paddingHorizontal": 0,
    "paddingTop": 24,
  },
  "header": {
    "alignItems": "center",
    "backgroundColor": "#FFFFFF",
    "flexDirection": "row",
    "minHeight": 64,
    "paddingBottom": 20,
    "paddingHorizontal": 20,
    "paddingTop": 16,
  },
  "keyboardView": {
    "flex": 1,
  },
  "submitButton": {
    "borderRadius": 12,
    "minHeight": 48,
    "width": "100%",
  },
  "submitButtonDisabled": {
    "backgroundColor": "#D0E0E3",
  },
  "submitButtonEnabled": {
    "backgroundColor": "#F39200",
  },
  "title": {
    "color": "#005CA9",
    "flex": 1,
    "fontSize": 18,
    "fontWeight": "600",
    "lineHeight": 24,
  },
}
`;
